ledger/src/blockstore.rs

```diff
@@ -2872,28 +2872,29 @@ impl Blockstore {
                     .and_then(|serialized_shred| {
                         if serialized_shred.is_none() {
                             if let Some(slot_meta) = slot_meta {
-                                panic!(
-                                    "Shred with
-                                    slot: {},
-                                    index: {},
-                                    consumed: {},
-                                    completed_indexes: {:?}
-                                    must exist if shred index was included in a range: {} {}",
-                                    slot,
-                                    i,
-                                    slot_meta.consumed,
-                                    slot_meta.completed_data_indexes,
-                                    start_index,
-                                    end_index
-                                );
-                            } else {
-                                return Err(BlockstoreError::InvalidShredData(Box::new(
-                                    bincode::ErrorKind::Custom(format!(
-                                        "Missing shred for slot {}, index {}",
-                                        slot, i
-                                    )),
-                                )));
+                                if slot >= self.lowest_cleanup_slot() {
```

Comment:
I think it should be slot > self.lowest_cleanup_slot(). The lowest_cleanup_slot might have been cleaned up.

Reply:
See 

    /// Performs cleanup based on the specified deletion range.  After this
    /// function call, entries within \[`from_slot`, `to_slot`\] will become
    /// unavailable to the reader immediately, while its disk space occupied
    /// by the deletion entries are reclaimed later via RocksDB's background
    /// compaction.
    ///
    /// Note that this function modifies multiple column families at the same
    /// time and might break the consistency between different column families
    /// as it does not update the associated slot-meta entries that refer to
    /// the deleted entries.
    ///
    /// For slot-id based column families, the purge is done by range deletion,
    /// while the non-slot-id based column families, `cf::TransactionStatus`,
    /// `AddressSignature`, and `cf::TransactionStatusIndex`, are cleaned-up
    /// based on the `purge_type` setting.
    pub fn purge_slots(&self, from_slot: Slot, to_slot: Slot, purge_type: PurgeType) {

Reply:
Yep, you're right.  I'm also checking rocksdb's document and also found the inconsistency between its DeleteRange and DeleteFilesInRange and thus publish https://github.com/solana-labs/solana/pull/27201.

Reply:
Aghh, our code comments aren't consistent.  I think the correct meaning of the `lowest_cleanup_slot` is the oldest slot that is still available.  Let me file a different PR to clean up this mess.

in [ledger_cleanup_service.rs](https://github.com/yhchiang-sol/solana/blob/master/core/src/ledger_cleanup_service.rs#L251-L252):
```
                    // purge any slots older than lowest_cleanup_slot.
                    blockstore.purge_slots(0, lowest_cleanup_slot, PurgeType::CompactionFilter);
```

And in [blockstore.rs](https://github.com/yhchiang-sol/solana/blob/master/ledger/src/blockstore.rs#L2184-L2185) where we compare `lowest_cleanup_slot`
```
        // lowest_cleanup_slot is the last slot that was not cleaned up by LedgerCleanupService
        let lowest_cleanup_slot = self.lowest_cleanup_slot.read().unwrap();
```

Reply:
Offline discussed with @lijunwangs, looks like `lowest_cleanup_slot` does have correct meaning as we perform `[from, to + 1)` inside our purge logic.

```
    /// A helper function to `purge_slots` that executes the ledger clean up
    /// from `from_slot` to `to_slot`.
    pub(crate) fn run_purge_with_stats(
        &self,
        from_slot: Slot,
        to_slot: Slot,
        purge_type: PurgeType,
        purge_stats: &mut PurgeStats,
    ) -> Result<bool> {
        let mut write_batch = self
            .db
            .batch()
            .expect("Database Error: Failed to get write batch");
        // delete range cf is not inclusive
        let to_slot = to_slot.saturating_add(1);
```

So it should be the logic that I mentioned above in [blockstore.rs](https://github.com/yhchiang-sol/solana/blob/master/ledger/src/blockstore.rs#L2184-L2185) that needs to be fixed instead:

```
        // lowest_cleanup_slot is the last slot that was not cleaned up by LedgerCleanupService
        let lowest_cleanup_slot = self.lowest_cleanup_slot.read().unwrap();
```

Reply:
(haha, yeah, those off-by-one (inclusive/exclusive) thing was hard to do right back when i worked on the blockstore code.. )